File: TeamCode/src/main/java/org/firstinspires/ftc/common/opmode/BrickOpMode.java
package org.firstinspires.ftc.common.opmode;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.util.ElapsedTime;

import org.firstinspires.ftc.common.commandbase.CommandScheduler;
import org.firstinspires.ftc.common.hardware.subsystems.SubsystemManager;
import org.firstinspires.ftc.common.hardware.subsystems.Updatable;
import org.firstinspires.ftc.common.logicnodes.NodeManager;

import java.util.ArrayList;
import java.util.List;

public abstract class BrickOpMode extends LinearOpMode {
    public enum OpModeType {
        TELE_OP,
        AUTONOMOUS
    }

    public static CommandScheduler commandScheduler = CommandScheduler.getInstance();
    public static NodeManager nodeManager = NodeManager.getInstance();
    public static SubsystemManager subsystemManager = SubsystemManager.getInstance();

    public static List<Updatable> updatableList = new ArrayList<>();

    private static final ElapsedTime runtime = new ElapsedTime();

    // TODO: TEST THIS AS IT PROBABLY WON'T WORK
    public OpModeType opModeType = this.getClass().isAnnotationPresent(Autonomous.class) ? OpModeType.AUTONOMOUS : OpModeType.TELE_OP;

    private long loopCount = 0;
    private double loopFrequency;

    /**
     * This method is called once when the driver hits INIT
     */
    public abstract void onInit();

    /**
     * This method is called repeatedly after driver hits INIT, but before they hit START
     */
    public abstract void initLoop();

    /**
     * This method is called once when the driver hits START
     */
    public abstract void onStart();

    /**
     * This method is called repeatedly after driver hits START
     */
    public abstract void run();

    @Override
    public void runOpMode() {
        commandScheduler.reset();
        subsystemManager.clear();
        nodeManager.clear();

        onInit();
        while (!isStarted() && !isStopRequested()) {
            initLoop();
        }
        waitForStart();

        runtime.reset();
        onStart();

        while (opModeIsActive() && !isStopRequested()) {
            // This runs the autonomous nodes logic
            if (opModeType == OpModeType.AUTONOMOUS) {
                nodeManager.run();
            }

            // This is the user's run method
            run();

            // This runs the command queue
            commandScheduler.run();

            // This runs the read, compute, write methods of the subsystems
            subsystemManager.run();

            // This runs the update method for each updatable object
            updatableList.forEach(Updatable::update);

            loopFrequency = ++loopCount / runtime.seconds();
        }
    }
}

File: TeamCode/src/main/java/brickbot/quickstart/opmode/BrickOpMode.kt
package brickbot.quickstart.opmode

import brickbot.quickstart.commandbase.CommandScheduler
import brickbot.quickstart.devices.DeviceManager
import brickbot.quickstart.subsystems.SubsystemManager
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.util.ElapsedTime

abstract class BrickOpMode: LinearOpMode() {
    protected val commandScheduler = CommandScheduler
    protected val subsystemManager = SubsystemManager

    private val deviceManager = DeviceManager
    private val runtime = ElapsedTime()

    private var loopCount = 0
    protected var loopFrequency: Double = 0.0

    /**
     * This method is called once when the driver hits INIT.
     */
    abstract fun onInit()

    /**
     * This method is called repeatedly after driver hits INIT, but before they hit START.
     */
    abstract fun initLoop()

    /**
     * This method is called once when the driver hits START.
     */
    abstract fun onStart()

    /**
     * This method is called repeatedly after driver hits START.
     */
    abstract fun run()

    override fun runOpMode() {
        commandScheduler.reset()
        subsystemManager.clear()

        // This calls the init methods automatically for all devices
        deviceManager.initDevices()

        onInit()
        while (!isStarted() && !isStopRequested()) {
            initLoop()
        }
        waitForStart()

        runtime.reset()
        onStart()

        while (opModeIsActive() && !isStopRequested()) {
            // This is the user's run method
            run()

            // This runs the command queue
            commandScheduler.run()

            // This runs the read, compute, write methods of the subsystems
            subsystemManager.run()

            loopFrequency = ++loopCount / runtime.seconds()
        }
    }
}